### DISCUSSION TITLE
Running Closure Compiler without Rebuilding

### DISCUSSION THREAD

**Discussion by takase1121** [emscripten-core/emscripten#22039] [Q&A] [answered]
I am working on porting Lite XL over to run on browsers. I have JS files that provide better integrations that are bundled by ESBuild. I want to use Closure because it reduces the minified JS size by quite a lot (IIRC about half the original) which is more than what ESBuild can do at least.

I have some work that allows me to mix ESBuild and Emscripten's Closure compiler somehow, and it works by bundling with ESBuild and injecting that via emcc's `--pre-js` option (I'm jumping between meson and npm for all these build steps and it's messy). The compilation takes quite a long time, but with Closure it takes even longer.

Is there a way to use Closure without having to rebuild (relink) the program by just taking the intermediate emcc output and running it against Closure? I tried running the output from emcc (without Closure) through `closure-compiler -O ADVANCED` but that didn't work because emcc has a bunch of extensions and preprocessing steps. That should significantly cut down the build time when I only needed to change the JS file, not the C sources.

For your reference, here is the time for non-Closure builds:
![image](https://github.com/emscripten-core/emscripten/assets/20792268/d79ea418-fe84-442e-829c-fefcb66a77e6)

Here is the time for Closure builds:
![image](https://github.com/emscripten-core/emscripten/assets/20792268/0b043d6a-3137-424d-90d6-a54679d11d72)

Closure seems to take 6 seconds to run, but it's better than 16 seconds for a 1-line change in JS.

**Comment by kripken** [comment, marked as answer]
You can link without full minification using `--minify=0`, which should then allow closure advanced to run properly after Emscripten. Here is the test for that:

https://github.com/emscripten-core/emscripten/blob/a6e1f5c0792393059510430831d908f74e13e1ce/test/test_other.py#L14567-L14574

When full minification is on (which it is by default in `-O2` and above) then we modify the JS in ways that require integration with closure (which we do when we run closure ourselves, but if you run it yourself, you need to disable minification).

**Reply by takase1121** [reply]
Looks like there are more stuff that needs to be considered like closure-externs but this is good enough for me. Thank you!

**Reply by sbc100** [reply]
Is there some reason why ESBuild doesn't work with the output emscripten when using the internal closure compiler?  The internal use of closure compiler should only ever minify internal stuff.. so maybe something is going wrong there that it would be better to fix here in emscripten?

**Reply by takase1121** [reply]
No, it's not a problem with emscripten here per-se, my shell files are compiled with esbuild and included into the output wasm+js file combo with `--pre-js`. This however requires me to relink the program whenever my shell file changes, which is painful with `-O3` and asyncify going on. My shell files also depend on some FS API functions, so I suspect there will be issues calling the API on the closure-compiled output.

I want emscripten to emit the original file before modification, do all the bundling with esbuild and then run that file over closure compiler.

As a sidenote, is there a good way to get the errno code constants?

**Reply by sbc100** [reply]
For errno constants, you have a few options:

1. From withing JS library code (or `--pre-js`/`--post-js` code if you add `#preprocess` to the top of the file) you can use a macro (e.g. `{{{ cDefs.EPERM }}}`.
2. From anywhere you get the string version of the code by calling the native `_strerror(code)` function.
3. You can depend on (or export) the `ERRNO_CODES` JS mapping here: https://github.com/emscripten-core/emscripten/blob/3de1b9091adbadc46706959bd97efa033214bc2c/src/library.js#L1309-L1312.
